### ISSUE TITLE
Should .git folder be downloaded with Composer?

### ISSUE THREAD

**Issue by RainbowArray** [Lullabot/amp-library#30]
When I installed this via Composer and uploaded the files to a server, I received an error about a git submodule being within my repo. I looked into the amp library files in the repo and found a .git directory within the amp php library directory, which seemed unlike the other libraries in my vendors directory. Once I removed the .git directory, I was able to upload the library to the server without an issue. Not sure if something should be changed in the setup, but wanted to note this.

**Comment by sidkshatriya** [comment]
This is not a bug. The module `composer.json` has 

``` javascript
"repositories": [
    {
      "type": "git",
      "url": "https://github.com/Lullabot/amp-library"
    }
  ],
  "require": {
    "lullabot/amp": "dev-master"
  }
```

What that means is that composer will get the amp-library source from git (github) instead of packagist (this workflow had originally been suggested by @karens ). This what allowed us to use composer without putting it up on packagist. The other vendor folders without .git are hosted by packagist.

Have we guys got a Lullabot account on packagist (like we do on github) ? Karen mentioned she has got a account available -- she can now set it up there if she wants.

The current setup works fine and I'm wondering if there are any disadvantages in continuing as we do currently. This setup allows us to do neat things like `composer update lullabot/amp` and then `vendor/lullabot/amp` folder updates to the latest commit in git.

@karens can proceed as she feels comfortable 

cc: @sirkitree

**Comment by sidkshatriya** [comment]
If we're setting this up on packagist we should also consider setting a github hook as suggested here: https://packagist.org/about#how-to-update-packages 

We also don't seem to lose the ability to simply track a branch instead of a version tag ( https://packagist.org/about )

> Branches will automatically appear as "dev" versions that are easily installable by anyone that wants to try your library's latest and greatest, but that does not mean you should not tag releases.

**Comment by sidkshatriya** [comment]
Another point, if we're going with packagist we should tag the AMP library as something like 1.0.0-beta1 perhaps?

The `composer.json` in D7 and D8 modules will now need to be changed to remove the snippet above and require 1.0.0-beta1

**Comment by ghost** [comment]
Packagist is set up now, https://packagist.org/packages/lullabot/amp. I also set up the web service to automatically update packagist. We could update the module to retrieve the library from packagist now. If so we may want to create a tag.

**Comment by ghost** [comment]
We could also just track the branch instead of a tag, might be easier with all the changes

**Comment by RainbowArray** [comment]
I think branch sounds easier than a tag. Otherwise we will need to make lots and lots of modules to keep up with the library changes.

Thanks for setting this up on Packagist!

**Comment by sidkshatriya** [comment]
@karens Thanks for creating the package on packagist!

@mdrummond Yes, I agree that we should track the master branch in our D7 and D8 modules. But this does _not_ solve our .git folder problem:

As an experiment for the D8 module I did:

``` diff
--- a/composer.json
+++ b/composer.json
@@ -10,12 +10,6 @@
     "issues": "http://drupal.org/project/issues/amp",
     "source": "http://cgit.drupalcode.org/amp"
   },
-  "repositories": [
-    {
-      "type": "git",
-      "url": "https://github.com/Lullabot/amp-library"
-    }
-  ],
   "require": {
     "lullabot/amp": "dev-master"
   }
```

Then I proceed to nuke my composer cache, composer lock file, vendor directory to make sure that I have a fresh slate. Doing `composer install` _still_ results in a .git folder being added as we're requiring a version that is tracking a branch.

Composer documentation confirms this:
https://getcomposer.org/doc/02-libraries.md#branches

> Note: When you install a development version, it will be automatically pulled from its source. See the install command for more details.

But there is a way to avoid the creation of a .git folder . And that is to use
`composer install --prefer-dist`

See http://stackoverflow.com/questions/25981292/how-to-make-composer-not-create-a-git-directory-for-a-package
also:
https://getcomposer.org/doc/03-cli.md#install

When I did this, a .git folder was _not_ created

As a further experiment, I created a tag `1.0.0-beta1` and changed the composer.json to:

```
   "require": {
     "lullabot/amp": "~1.0.0-beta1"
   }
```

This time a .git folder, expectedly was _not_ created during `composer install`

Next I wondered if we can somehow avoid constantly changing `composer.json` as @mdrummond feared. There is a way out. And that is to do:

```
   "require": {
     "lullabot/amp": "1.0.*"
   }
```

As as further experiment, I created a new tag `1.0.0-beta2`, With the above version constraint, we are able to _successfully_ download the new version. (See http://qpleple.com/understand-composer-versions/ for the various ways we can name versions)

**So in summary:**
I suggest we could continue tracking dev-master in require (and remove the repositories section). If people are fussy about the .git folder they can use the --prefer-dist command line option. 

The alternative approach is to require `1.0.*` but the slight disadvantage of that is that we need to keep creating tags on the amp-library. Note that I only need to change the composer.json _once_ for the D7 and D8 modules.

@mdrummond @karens Let me know if we want to `require: "1.0.*"` (needs constant tagging in amp-library) or `require:"dev-master"` (creates .git folder unless invoked with --prefer-dist)

**Comment by sidkshatriya** [comment]
There is another advantage with the "1.0.*" approach in that you can tag known "good" versions and keep updating the master branch without too much worry.

**Comment by RainbowArray** [comment]
I think it probably makes sense to start tagging versions for the library.
That seems more akin to what other projects do. It seems like a pretty
common workflow to put a certain range of versions into composer. Needing
to use prefer-dist command feels like a kludge. If this was earlier in the
development cycle that might make sense, but I think the library is a lot
more stable now, so versions seem like a good way to go.

On Tue, Mar 29, 2016 at 1:46 PM, Sidharth Kshatriya <
notifications@github.com> wrote:

> There is another advantage with the "1.0.*" approach in that you can tag
> known "good" versions and keep updating the master branch without too much
> tension.
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> https://github.com/Lullabot/amp-library/issues/30#issuecomment-203044975

## 

Marc Drummond | Front-end developer | Lullabot
e: marc@lullabot.com w: www.lullabot.com t: twitter.com/MarcDrummond

**Comment by sidkshatriya** [comment]
@karens and I had a discussion on slack. We share your point of view. Will update D7 and D8 `composer.json` now to require `1.0.*` and remove the `repositories` section in composer.json

**Comment by sidkshatriya** [comment]
This is complete:
- Updated D7 and D8 module composer.json
  http://drupalcode.org/project/amp.git/commit/55efec0
  http://drupalcode.org/project/amp.git/commit/ccb0309
- Updated README.md
